File: 2022/day_01.py
"""Day 1 AoC 2022"""
from itertools import groupby

if __name__ == "__main__":

    with open("./day_01_input.txt") as puzzle_input:
        all_elves_calories = [
            # convert all the string numbers to ints
            list(map(int, group))
            # break up the lists by where they are empty or not with bool(item_to_check)
            for key, group in groupby([line.strip() for line in puzzle_input], key=bool)
            # if the group is empty or only one blank, skip it.
            if key
        ]

    total_calories_by_elf = sorted(
        [sum(elf_calories) for elf_calories in all_elves_calories], reverse=True
    )
    print(f"The elf with the most calories has {max(total_calories_by_elf)} calories.")
    print(f"The top three elves have a total of {sum(total_calories_by_elf[0:3])} calories.")

File: 2022/day_01.cpp
#include <algorithm>
#include <numeric>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace std;

int main(void)
{
    vector<int> elves;
    int current_calories = 0;

    ifstream infile("day_01_input.txt");
    std::string line_in_file;
    while (std::getline(infile, line_in_file))
    {
        if (line_in_file.empty())
        {
            elves.push_back(current_calories);
            current_calories = 0;
        }
        else
        {

            current_calories += std::stoi(line_in_file);
        }
    }

    std::sort(elves.begin(), elves.end());
    cout << "Part 1: Elf with the most calories: " << elves.back() << endl;

    // Let it infer the op and just sum ints.
    cout << "Part 2: Top Three Elves: " << std::accumulate(elves.end() - 3, elves.end(), 0) << endl;
    return 0;
}
